### ISSUE TITLE
About Query Criteria Time Interval

### ISSUE THREAD

**Issue by yangshike** [quickwit-oss/quickwit#3165]
I want to query data for a given time interval, but I don't know how to use this query statement
my config：
  - name: msgTime
      type: datetime
      fast: true
      input_formats:
        - rfc3339
        - "%Y-%m-%d %H:%M:%S"
  timestamp_field: msgTime

Such as  i want to query  msgTime  between '2023-01-01 00:00:00' and '2023-01-31 23:59:59'

curl "http://127.0.0.1:7280/api/v1/wework/search?query=msgContent:sss+AND+msgTime:??'"

**Comment by fmassot** [comment]
@yangshike you should use `start_timestamp` and `end_timestamp` query parameters that are expressed in seconds. More details in the docs: https://quickwit.io/docs/reference/rest-api#parameters-1

**Comment by fmassot** [comment]
You don't need to change the format.

Your curl should be like this:

```bash
curl "http://127.0.0.1:7280/api/v1/wework/search?query=*&start_timestamp=1681154472"
```

**Comment by fmassot** [comment]
@evanxg852000 I think we need a guide on the timestamp field pruning.

**Comment by PSeitz** [comment]
The date format to query should be rfc3339. The queryparser can handle ranges in the format `msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z}`

Inclusive bounds are [], exclusive are {}

Range queries need to have a fast field (which is the case here)

**Comment by evanxg852000** [comment]
I would just like to add that for the internal storage of the value, they all end up in timestamps, and the detail are [here](https://quickwit.io/docs/configuration/index-config#datetime-type)
- `start_timestamp` and `end_timestamp` are not part of your index fields, these are metadata fields to help pruning split. They cannot be used in the query as fields.
- The datetime field you can actually use in your query is `msgTime` and the only format supported in the query parser as of now is `rfc3339`

**Comment by yangshike** [comment]
> The date format to query should be rfc3339. The queryparser can handle ranges in the format `msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z}`
> 
> Inclusive bounds are [], exclusive are {}
> 
> Range queries need to have a fast field (which is the case here)

 curl "http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z]"
curl: (3) bad range in URL position 64:
http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z]

**Comment by yangshike** [comment]
> I would just like to add that for the internal storage of the value, they all end up in timestamps, and the detail are [here](https://quickwit.io/docs/configuration/index-config#datetime-type)
> 
> * `start_timestamp` and `end_timestamp` are not part of your index fields, these are metadata fields to help pruning split. They cannot be used in the query as fields.
> * The datetime field you can actually use in your query is `msgTime` and the only format supported in the query parser as of now is `rfc3339`
> 
> @yangshike we will add a guide on this and possibly the limitation on query parser supporting only one format when querying will be lifted when the new query pipeline lands.

The pruning of time intervals is very important. I would like to know how to prune by time intervals when generating index data, or if adding time intervals during my query can reduce the amount of scanned data

**Comment by yangshike** [comment]
> > The date format to query should be rfc3339. The queryparser can handle ranges in the format `msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z}`
> > Inclusive bounds are [], exclusive are {}
> > Range queries need to have a fast field (which is the case here)
> 
> curl "[http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:[2002-10-02T15:00:00Z](http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:%5B2002-10-02T15:00:00Z) TO 2002-10-02T18:00:00Z]" curl: (3) bad range in URL position 64: [http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:[2002-10-02T15:00:00Z](http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:%5B2002-10-02T15:00:00Z) TO 2002-10-02T18:00:00Z]
the documents:
https://quickwit.io/docs/reference/query-language
:
Range queries
Range queries can only be executed on fields with a fast field. Currently only fields of type ip are supported.

**Comment by PSeitz** [comment]
> > The date format to query should be rfc3339. The queryparser can handle ranges in the format `msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z}`
> > Inclusive bounds are [], exclusive are {}
> > Range queries need to have a fast field (which is the case here)
> 
> curl "[http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:[2002-10-02T15:00:00Z](http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:%5B2002-10-02T15:00:00Z) TO 2002-10-02T18:00:00Z]" curl: (3) bad range in URL position 64: [http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:[2002-10-02T15:00:00Z](http://127.0.0.1:7280/api/v1/wework/search?query=msgTime:%5B2002-10-02T15:00:00Z) TO 2002-10-02T18:00:00Z]

The query needs to be properly escaped. JSON as query body is easier to avoid curl escaping:

`curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z]"}' http://127.0.0.1:7280/api/v1/wework/search`

**Comment by yangshike** [comment]
I have another question.
 curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z] AND msgContent:yyyy"}' http://127.0.0.1:7280/api/v1/wework/search
curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgContent:yyyy"}' http://127.0.0.1:7280/api/v1/wework/search

These two queries consume almost the same amount of time,

Is time trimming ineffective in improving query speed?

**Comment by evanxg852000** [comment]
> The pruning of time intervals is very important. I would like to know how to prune by time intervals when generating index data, or if adding time intervals during my query can reduce the amount of scanned data
Yes the time pruning that is designed to reduces the amount of scanned data (split) is the one used on `start_timestamp` and `end_timestamp`

**Comment by yangshike** [comment]
> curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z] AND msgContent:yyyy"}' http://127.0.0.1:7280/api/v1/wework/search

Based on this query, how can i added  added start_timestamp and end_ timestamp？

**Comment by evanxg852000** [comment]
if my conversion is correct it should be:
2023-04-07T15:00:00Z  -> 1680879600s
2023-04-07T18:00:00Z -> 1680890400

`curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z] AND msgContent:yyyy"}' http://127.0.0.1:7280/api/v1/wework/search?start_timestamp=1680879600&end_timestamp=1680890400`

The time pruning on splits (start_timestamp and end_ timestamp) is applied first, then the search for msgTime is done.

**Comment by PSeitz** [comment]
I would not mix post and get parameters, the query should be

curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z]", "start_timestamp":1680879600, "end_timestamp":1680890400}' http://127.0.0.1:7280/api/v1/wework/search

**Comment by yangshike** [comment]
> Note: you cannot just use msgTime:*, you still have to restrict your field msgTime since a split might contain a much larger range, therefore, documents that are outside of [2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z]

I think if my msgTime interval is the same as  start_timestamp and end_timestamp interval , so there is no need for msgTime for query criteria

**Comment by yangshike** [comment]
> I would not mix post and get parameters, the query should be: `curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2002-10-02T15:00:00Z TO 2002-10-02T18:00:00Z]", "start_timestamp":1680879600, "end_timestamp":1680890400}' http://127.0.0.1:7280/api/v1/wework/search`

thank you very much ,its work ok!

**Comment by yangshike** [comment]
Here comes a new problem!!

curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z] AND msgContent:yyyy","start_timestamp":1680850800,"end_timestamp":1680861600}' "http://127.0.0.1:7280/api/v1/wework/search"
the response:
  "num_hits": 0


curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-7T18:00:00Z] AND msgContent:yyyy"}' "http://127.0.0.1:7280/api/v1/wework/search"
the response:
{
  "num_hits": 2}



this is the msgTIme in my documents:
{
      "msgSeq": 1206588,
      "msgTime": "2023-04-07T15:13:20Z",
      "msgType": "text"
}
{
 "msgSeq": 1210152,
      "msgTime": "2023-04-07T17:54:59Z",
      "msgType": "text"
}

**Comment by fulmicoton** [comment]
@yangshike just to be sure: can you join your index config? You need to explicitly declare the msgTime field as your timestamp field in the index config.

**Comment by PSeitz** [comment]
end_timestamp 1680861600 = Apr 07 2023 10:00:00
So no results is correct. There is no timezone or UTC or GMT+0.

I used this tool: https://www.epochconverter.com/

**Comment by yangshike** [comment]
The time zone should be related to the environment variables of the operating system. The time I checked on the operating system is normal
The time zone should be related to the environment variables of the operating system, is it normal for me to query this time on the operating system?

Do I need to set the time zone for quwickwit?

clickhouse01 :) select toDateTime(1680861600);

SELECT toDateTime(1680861600)

Query id: e897094b-b9ba-4eff-bfa5-fa3114040401
┌─toDateTime(1680861600)─┐
│    2023-04-07 18:00:00 │
└────────────────────────┘

1 row in set. Elapsed: 0.001 sec.

**Comment by PSeitz** [comment]
There's not timezone handling in quickwit yet, so it will not take the timezone of your OS. Queries need to be expressed in UTC/GMT+0 timezone.

**Comment by yangshike** [comment]
I have temporarily resolved this issue because there is a timestamp field in my source data, and I have set this field to timestamp_ field: msgTimestamp

previous timestamp_field  in config file is  :  timestamp_ field: msgTime

but my msgTime  like this : "msgTime": "2023-04-07T17:54:59Z",

my config file ：

  - name: msgTime
      type: datetime
      input_formats:
        - rfc3339
        - "%Y-%m-%d %H:%M:%S"
      fast: true
 
I think the problem is that Quickwit chose the wrong time zone when converting msgTime to timestamp

**Comment by PSeitz** [comment]
Since there's no timezone handling in quickwit, indexed data will be interpreted as UTC.

I'm not sure if the VRL feature can be used to normalize the indexed data to UTC ?
@guilload

**Comment by guilload** [comment]
It does not seem that there is anything wrong with parsing the field `msgTime` in Quickwit. `2023-04-07T17:54:59Z` is a valid RFC 3339 datetime. The suffix `Z` indicates that the timezone is UTC. I ran the unit tests for the RFC 3339 datetime parser, and it passed.

```rust
let date_time = parse_rfc3339("2023-04-07T17:54:59Z").unwrap();
assert_eq!(date_time, datetime!(2023-04-07 17:54:59 UTC));
 ```
 
I believe the issue is not passing UTC timestamps to Quickwit in the query. If I use your ClickHouse example, `2023-04-07 18:00:00` in RFC 3339 format (UTC) is `2023-04-07T18:00:00Z`. The corresponding timestamp is `1680890400`, not `1680861600`.
 
 ```rust
 let date_time = parse_rfc3339("2023-04-07T18:00:00Z").unwrap();
assert_eq!(date_time, datetime!(2023-04-07 18:00:00 UTC));
assert_eq!(date_time.unix_timestamp(), 1680890400);
```

**Comment by yangshike** [comment]
Okay, it could be that there was a problem with the source data conversion. My msgTime in kafka is as follows: msgTime: "2023-04-07 17:54:59"

But after saving into quwickwit, the msgTime will automatically change to: msgTime:"2023-04-07T17:54:59Z "

**Comment by guilload** [comment]
Ok, so in that case, it will use the second parser and assume the timezone is UTC. Unfortunately, at the moment, Quickwit does not support specifying a timezone manually.

The solution that @PSeitz mentioned is to add a [transform parameter](https://quickwit.io/docs/main-branch/configuration/source-config#transform-parameters) to your source to append the correct time zone.

```yaml
source_id: ...
source_type: kafka
...
transform:
  script: |
    .msgTime = join!([.msgTime, "+08:00"], separator: " ")
```

Under the hood, it is using [VRL](https://vector.dev/docs/reference/vrl/).

Don't forget to modifier the datetime parser format to `"%Y-%m-%d %H:%M:%S %z"`.

**Comment by yangshike** [comment]
Okay, thank you. This issue has been temporarily resolved,

There should be a certain performance loss through transform script, right,

However, it would be best if you could manually specify the time zone on quwickwit.

**Comment by yangshike** [comment]
> if my conversion is correct it should be: 2023-04-07T15:00:00Z -> 1680879600 seconds 2023-04-07T18:00:00Z -> 1680890400 seconds
> 
> `curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z] AND msgContent:yyyy"}' http://127.0.0.1:7280/api/v1/wework/search?start_timestamp=1680879600&end_timestamp=1680890400`
> 
> The time pruning on splits (start_timestamp and end_ timestamp) is applied first, then the search for msgTime is done.
> 
> **Note**: you cannot just use `msgTime:*`, you still have to restrict your field `msgTime` since a split might contain a much larger range, therefore, documents that are outside of [2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z]

I tested it, and if only use start_timestamp and end_timestamp, The performance of is twice for 

Add start _timestamp and end_timestamp and msgTime: twice the ["xx" TO 'xx']

**Comment by fulmicoton** [comment]
That's correct. We do not handle pruning when the range is supplied in the user query

**Comment by yangshike** [comment]
curl -X POST -H 'Content-Type: application/json' -d '{"query": "msgTime:[2023-04-07T15:00:00Z TO 2023-04-07T18:00:00Z] AND msgContent:yyyy","start_timestamp":168080800,"end_timestamp":1680861600}' "http://127.0.0.1:7280/api/v1/wework/search"
I also provided start_timestamp and end_ timestamp for this query。 Shouldn't we cut it first?

**Comment by fulmicoton** [comment]
Ah yes then pruning must have kicked in.

Post-pruning, the way the range is enforced is different in both case.

In the first query it is running as a Range query.
In the second query this is a "post filter" at collection level today.

It might be worth investigating which one is best an make sure the best implementation kicks where it makes sense..
